In [11]:
import pandas as pd
df=pd.read_excel('Exercise1(RA).xlsx')

In [12]:
df.head()

education_years  working_years  salary
0                6              7      38
1               12             14      53
2                9              5      40
3               16              8      50
4               18              6      55

# 一、線性迴歸模式：解釋分析模式
用於分析哪些自變數(因素)會影響到某個依變數之統計分析模式，其中依變數需為“屬量”的變數，自變數亦為屬量變數

    可參考https://www.statsmodels.org/stable/api.html

In [13]:
X=df[["education_years", "working_years"]] #自變數
Y=df["salary"]

In [24]:
#語法一
import statsmodels.api as smf
XX=smf.add_constant(X) #截句
RA=smf.OLS(Y,XX).fit()
print(RA.summary())

                            OLS Regression Results                            
Dep. Variable:                 salary   R-squared:                       0.914
Model:                            OLS   Adj. R-squared:                  0.905
Method:                 Least Squares   F-statistic:                     101.6
Date:                Thu, 12 May 2022   Prob (F-statistic):           7.16e-11
Time:                        08:15:13   Log-Likelihood:                -42.477
No. Observations:                  22   AIC:                             90.95
Df Residuals:                      19   BIC:                             94.23
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
const              23.6193      1.835     

/usr/local/lib/python3.7/dist-packages/statsmodels/tsa/tsatools.py:117: FutureWarning: In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only
  x = pd.concat(x[::order], 1)


F-statistic(F值):101.6

Prob(P值):7.16e-11(趨近於0，小於0.05有顯著相關)  

R-squared:0.914(解釋能力很好，因為趨近於1)

education_years的T值大於working_years，代表更有顯著相關

In [15]:
#語法一
#進行迴歸分析，跑出迴歸分析解釋分析表(從statsmodels.formula.api 匯入ols線性迴歸分析)
#PTS為依變數；A, OEFF,TO為自變數，進行迴歸分析
import statsmodels.formula.api as smf
model = smf.ols("salary ~ education_years+ working_years ", df).fit()
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                 salary   R-squared:                       0.914
Model:                            OLS   Adj. R-squared:                  0.905
Method:                 Least Squares   F-statistic:                     101.6
Date:                Thu, 12 May 2022   Prob (F-statistic):           7.16e-11
Time:                        07:23:11   Log-Likelihood:                -42.477
No. Observations:                  22   AIC:                             90.95
Df Residuals:                      19   BIC:                             94.23
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
Intercept          23.6193      1.835     

In [16]:
df.head()

education_years  working_years  salary
0                6              7      38
1               12             14      53
2                9              5      40
3               16              8      50
4               18              6      55

# 二、線性迴歸模式：預測分析模式_機器學習
利用線性迴歸模型，建立依變數之預測模型

In [17]:
#2.用線性回歸做訓練線性模型:
#import基本函式庫
import pandas as pd
import numpy as np

#準備X & y array
X = df[["education_years","working_years"]] #education_years, working_years為自變數,可自行調整
y = df['salary']            #salary為依變數,可自行調整

#將資料分成訓練組及測試組
from sklearn.model_selection import train_test_split

#test_size代表測試組比例。random_state代表設定隨機種子，讓測試結果可被重複
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.25, random_state=3)

#載入線性迴歸，並訓練模型
from sklearn.linear_model import LinearRegression
lm = LinearRegression()
lm.fit(X_train,y_train)

#取得截距。如果公式是y=a+bx，a即是截距
print(lm.intercept_)
print(lm.coef_)

#取得迴歸係數，並用Data Frame顯示:pd.DataFrame(資料,列的名稱,行的名稱)
#lm.coef_
#X_train.columns
cdf = pd.DataFrame(lm.coef_,X_train.columns,columns=['Coef'])
cdf

25.538882436572624
[1.40497907 0.64603316]


Coef
education_years  1.404979
working_years    0.646033

計算模式績效:RMSE, R_Square

In [18]:
#計算R_Square,
R2_train = lm.score(X_train,y_train)
R2_test = lm.score(X_test,y_test)

print('train_R2=%.3f, test_R2=%.3f'%(R2_train,R2_test))

train_R2=0.914, test_R2=0.871


In [19]:
from sklearn.metrics import mean_squared_error
from math import sqrt
# 使用X_test預測y，得到y_pred1 
y_pred1 = lm.predict(X_test)
rmse = sqrt(mean_squared_error(y_test, y_pred1)) 
print("rmse: %.3f" % rmse) 

rmse: 2.059


In [20]:
df['salary'].mean()

46.0

In [21]:
#輸入X1,X2, 計算模式所得的y值
lm.predict([[12,15],[8,10]])

/usr/local/lib/python3.7/dist-packages/sklearn/base.py:451: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  "X does not have valid feature names, but"


array([52.08912874, 43.23904664])

In [22]:
#以陣列表示待預測之自變數
X_1=pd.DataFrame([[12,15],[8,10]])
print(X_1)

    0   1
0  12  15
1   8  10


In [23]:
#計算出上述陣列中自變數之預測值(y)
list(lm.predict(X_1))

/usr/local/lib/python3.7/dist-packages/sklearn/base.py:451: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  "X does not have valid feature names, but"


[52.08912873705921, 43.239046636897015]

# 練習題：請以nba2.xlsx之資料，以1Q＋2Q＋3Q, A, TOT, TO等因子建立PTS之線性迴歸模式，並說明分析結果。

In [31]:
df2=pd.read_excel('nba2.xlsx')
df2['TQ']=df2['1Q']+df2['2Q']+df2['3Q']

In [32]:
X=df2[["TQ",'A','TOT','TO']] #自變數
Y=df2["PTS"]

In [33]:
import statsmodels.api as smf
XX=smf.add_constant(X) #截句
RA=smf.OLS(Y,XX).fit()
print(RA.summary())

                            OLS Regression Results                            
Dep. Variable:                    PTS   R-squared:                       0.737
Model:                            OLS   Adj. R-squared:                  0.736
Method:                 Least Squares   F-statistic:                     1717.
Date:                Thu, 12 May 2022   Prob (F-statistic):               0.00
Time:                        08:31:20   Log-Likelihood:                -7991.8
No. Observations:                2460   AIC:                         1.599e+04
Df Residuals:                    2455   BIC:                         1.602e+04
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         24.4839      1.326     18.463      0.0

/usr/local/lib/python3.7/dist-packages/statsmodels/tsa/tsatools.py:117: FutureWarning: In a future version of pandas all arguments of concat except for the argument 'objs' will be keyword-only
  x = pd.concat(x[::order], 1)


F-statistic(F值):1717

Prob(P值):0(小於0.05有顯著相關)

R-squared:0.737(解釋能力好，因為趨近於1)

TQ:1Q+2Q+3Q的總和

T值的絕對值排序:TQ>A>TOT>TO，代表1Q+2Q+3Q的總和比其他三個更有顯著差異，TO相對較低

# **作業：**
# 請利用宜蘭生活機能.xlsx檔案資料分析治安"Total_secu", 交通"Total_traf",醫療"Total_phar",教育"Total_scho"等資源對房價“Total_Price”的影響。